### DISCUSSION TITLE
Session not closed after the main application get stoped

### DISCUSSION THREAD

**Discussion by ganesh1306** [FreeOpcUa/python-opcua#1288] [Q&A]
i'm using the below code for connection and read tag values

    def OpcUaClientConnection(connectionParam):
    
    try
        url = connectionParam.get('URL')
        print("URL of Server ", url)             
        opcuaClientHandler = Client(url)         
        
         
        opcuaClientHandler.set_user(connectionParam.get('username'))
            
        opcuaClientHandler.set_password(connectionParam.get('Password'))
        
        
        # like sprintf    
        Securitystring = "%s,%s,%s,%s" % (connectionParam.get('SecurityPolicy'), 
                                         connectionParam.get('SecurityMode'),
                                         connectionParam.get('certfile'),
                                         connectionParam.get('keyfile'))
        
        #Note : Call this before connect()                
        #print(Security string)
        opcuaClientHandler.set_security_string(Securitystring)                
        opcuaClientHandler.application_uri = "urn:freeopcua:client"
        
        connectionStatus = opcuaClientHandler.connect()
    
        
        if(connectionStatus == True):
            print("connection Establishment with server successful")
            print("Client connected to the URL", url)
            return True
        else:
            print("Failed to connect to server")
        
    except:        
        print("Connection Error!.")
        print("Server Disconnected Waiting for connection establishment")
        return False
    
    
def Disconnection():    
    
    try:
        print("disconnecting...")
        opcuaClientHandler.disconnect()     
          
    except:
        print("disconnection error!")      

this will run every one second when i stop the code on server side session not get closed

**Comment by AndreasHeine** [comment]
you wrapped your own class/function stuff around python-opcua so from my point of view you need to check/debug your code!
debug? breakpoints! incremental developement!

**Reply by ganesh1306** [reply]
can you suggest how do i unwrappe the function or class

**Reply by AndreasHeine** [reply]
that your job! i can only tell you that if you call client.disconnect() the session close request is send...

**Reply by ganesh1306** [reply]
i know  if i call client.disconnect() Session will be closed 
OpcUaClientConnection() is in while loop 
like
     while True:
       try:
         OpcUaClientConnection()
      except:
        Disconnection():
     sleep(1)

when forcefully stop the code (Server will be running) but session created on server side will not closed 

i know if server got stopped session will be closed by the above try and except condition 

but in node opc ua when i stop the client code by forcefully session will be closed on server 

i'm using prosys server

**Reply by AndreasHeine** [reply]
check sessiontimout on serverside... prosys has default 1h....

we send the close session request and what the server then did is not in our hands sorry!!!

**Reply by AndreasHeine** [reply]
why did you open a new session each sec anyways? sessions tend to be long lasting/living (longer then some seconds for cyclic reads) according to the opc ua specifications!

**Reply by Yashvanthb13** [reply]
> check sessiontimout on serverside... prosys has default 1h....
> 
> we send the close session request and what the server then did is not in our hands sorry!!!

pro sys as 1 hr timeout on server side

**Reply by Yashvanthb13** [reply]
> why did you open a new session each sec anyways? sessions tend to be long lasting/living (longer then some seconds for cyclic reads) according to the opc ua specifications!

once application/code is stopped by user the session will not be closed it take 1hr to close that session that has created when application/code got closed by user forcefully

**Reply by AndreasHeine** [reply]
means if you open a new session each second the server will delete the old ones after one hour and the limit of sessions serverside is reached pretty soon...

**Reply by ganesh1306** [reply]
yes right but i used node-opcua for testing when i close the application forcefully the session created on server side also get  closed 

	client_handler = OPCUAClient.create({

			keepSessionAlive: true,
			requestedSessionTimeout: 60000,
			securityMode: none,
			securityPolicy: node,
			requestedSessionTimeout: 10000,
			applicationName: "NodeOPCUA-Client",

			endpoint_must_exist: false,

			connectionStrategy: {
				maxRetry: 10000000,
				initialDelay: 100,
				maxDelay: 1000
			}
		});


above is the example of node-opcua session time 

One more if nodeID is not present on Server side, code getting crashing 

so i made little change in uatypes to return null if nodeid not available on server side because i was reading nodeid in loop if Nodeid   not present/the value is null instead or raising error we can return  null right

    def check(self):
        """
        Raises an exception if the status code is anything else than 0 (good).

        Use the is_good() method if you do not want an exception.
        """
        if not self.is_good():
            #raise UaStatusCodeError(self.value)

**Reply by AndreasHeine** [reply]
```
#raise UaStatusCodeError(self.value)
```
this is not a solution at all....

**Reply by ganesh1306** [reply]
then how can we solve this raise error when not able to read nodeid which are not available at server side so the application should not crash any suggestions on this

**Reply by ganesh1306** [reply]
> yes right but i used node-opcua for testing when i close the application forcefully the session created on server side also get closed
> 
> ```
> client_handler = OPCUAClient.create({
> 
> 		keepSessionAlive: true,
> 		requestedSessionTimeout: 60000,
> 		securityMode: none,
> 		securityPolicy: node,
> 		requestedSessionTimeout: 10000,
> 		applicationName: "NodeOPCUA-Client",
> 
> 		endpoint_must_exist: false,
> 
> 		connectionStrategy: {
> 			maxRetry: 10000000,
> 			initialDelay: 100,
> 			maxDelay: 1000
> 		}
> 	});
> ```
> 
> above is the example of node-opcua session time
>

For session closing issues what is your suggestions

**Reply by AndreasHeine** [reply]
cant confirm a issue...
![sessions](https://user-images.githubusercontent.com/56362817/111631594-dc6f0c80-87f3-11eb-8982-866579b12cef.PNG)

**Reply by ganesh1306** [reply]
Then need to wait till the session has to closed on server side

**Reply by AndreasHeine** [reply]
we already did....

https://github.com/FreeOpcUa/python-opcua/blob/bf734dfc00e96b3ba1c184c5f511589fa183166b/opcua/client/client.py#L292

https://github.com/FreeOpcUa/python-opcua/blob/bf734dfc00e96b3ba1c184c5f511589fa183166b/opcua/client/ua_client.py#L305

**Comment by AndreasHeine** [comment]
```
connect()
while 1:
    try:
        read
    except:
        stop = true
    if stop:
        break
    sleep
disconnect()
```
